### PR TITLE
MLDB-2113 run-script wait for all procs

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: server/mldb_runner.cc [hunk 1]
```diff
@@ -408,6 +408,15 @@ int main(int argc, char ** argv)
...
         if (!dontExitAfterScript) {
+            int tasksBefore = -1;
+            int tasks;
+            while ((tasks = BackgroundTaskBase::getRunningTasks()) != 0) {
+                if (tasks != tasksBefore) {
+                    tasksBefore = tasks;
+                    cerr << "Still waiting for " << tasks << " task(s)" << endl;
+                }
+                std::this_thread::sleep_for(std::chrono::milliseconds(100));
+            }
```
Comment (FinchPowers) [inline server/mldb_runner.cc]: Some tasks seem to never be destroyed so we now have locks!

File: rest/rest_collection.h [hunk 2]
```diff
@@ -117,7 +117,7 @@ struct BackgroundTaskBase {
     /** When destroying a background task, we make sure it's cancelled or
         finished so that there is nothing outstanding.
     */
-    ~BackgroundTaskBase();
+    virtual ~BackgroundTaskBase();
```
Comment (guyd) [inline rest/rest_collection.h]: Euh!  Ce destructeur fait quelque chose?  Il aurait du etre virtuel.

Comment (FinchPowers) [inline rest/rest_collection.h]: En effet! Scarry!

File: server/mldb_server.cc [hunk 3]
```diff
@@ -389,6 +390,17 @@ shutdown()
...
+    int tasksBefore = -1;
+    int tasks;
+    while ((tasks = BackgroundTaskBase::getPendingTasks()) != 0) {
+        if (tasks != tasksBefore) {
+            tasksBefore = tasks;
+            INFO_MSG(logger) << "Still waiting for " << tasks
+                             << " task(s)";
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
```
Comment (FinchPowers) [inline server/mldb_server.cc]: On hold. This is a band aid. Shutdown should shutdown and join services, thread pools and the like properly.

File: rest/rest_collection.cc [hunk 4]
```diff
@@ -648,11 +648,19 @@ BackgroundTaskBase::
 BackgroundTaskBase()
     : running(false), state(State::INITIALIZING)
 {
+    ++pendingTasks;
+    needToDecrementPendingTasks.store(1);
+}
+
+atomic<int> BackgroundTaskBase::pendingTasks{0};
+int BackgroundTaskBase::getPendingTasks() {
+    return BackgroundTaskBase::pendingTasks.load();
```
Comment (jeremybarnes) [inline rest/rest_collection.cc]: no need for load()

File: rest/rest_collection.cc [hunk 5]
```diff
@@ -648,11 +648,19 @@ BackgroundTaskBase::
 BackgroundTaskBase()
     : running(false), state(State::INITIALIZING)
 {
+    ++pendingTasks;
+    needToDecrementPendingTasks.store(1);
```
Comment (jeremybarnes) [inline rest/rest_collection.cc]: no need for store, use assignment

File: rest/rest_collection.h (hunk 2 above)
Comment (jeremybarnes) [inline rest/rest_collection.h]: It will only cause a problem if the derived class has a non-trivial destructor :)  but still scary

**Pull request closed without merging**